modyn/supervisor/internal/pipeline_executor/pipeline_executor.py

```diff
@@ -277,13 +279,33 @@ def _handle_triggers_within_batch(self, batch: list[tuple[int, int, int]], trigg
 
             num_samples_in_trigger = self.grpc.get_number_of_samples(self.pipeline_id, trigger_id)
             if num_samples_in_trigger > 0:
+                if i > 0:
+                    # since num_samples_in_trigger is not 0, we are sure that triggering_data is not empty
+                    first_timestamp = triggering_data[0][1]
+                    last_timestamp = triggering_data[-1][1]
+                else:
+                    # now it is the first trigger in this batch. Triggering_data can be empty.
+                    # when it is indeed empty, then there is remaining data in the last batch
+                    # because num_samples_in_trigger is not 0.
+                    assert len(triggering_data) > 0 or self.remaining_data_range is not None
+
+                    if self.remaining_data_range is not None:
+                        first_timestamp = self.remaining_data_range[0]
```

Comment:
What if `len(triggering_data)>0` in this case? Why not just use `triggering_data[0][1]`? I am a bit unsure why we need the remaining data range.

Assuming that `len(triggering_data)==0`, how can there be a first and last timestamp of the samples in the trigger?

Reply:
I answer it in the unit test Q & A!

Reply:
Thank you, I finally got it with this and the other comment!